dvc/rwlock.py

```diff
@@ -0,0 +1,182 @@
+from __future__ import unicode_literals
+
+import os
+import json
+
+from collections import defaultdict
+from contextlib import contextmanager
+
+from voluptuous import Schema, Invalid
+
+from .exceptions import DvcException
+from .lock import LockError
+from .utils.compat import (
+    convert_to_unicode,
+    FileNotFoundError,
+    JSONDecodeError,
+    str,
+)
+from .utils.fs import relpath
+
+INFO_SCHEMA = {"pid": int, "cmd": str}
+
+SCHEMA = Schema({"write": {str: INFO_SCHEMA}, "read": {str: [INFO_SCHEMA]}})
+
+
+class RWLockFileCorruptedError(DvcException):
+    def __init__(self, path, cause):
+        super(RWLockFileCorruptedError, self).__init__(
+            "Unable to read RWLock-file '{}'. JSON structure is "
+            "corrupted".format(relpath(path)),
+            cause=cause,
+        )
+
+
+class RWLockFileFormatError(DvcException):
+    def __init__(self, path, cause):
+        super(RWLockFileFormatError, self).__init__(
+            "RWLock-file '{}' format error.".format(relpath(path)), cause=cause
+        )
+
+
+@contextmanager
+def _edit_rwlock(lock_dir):
+    path = os.path.join(lock_dir, "rwlock")
+    try:
+        with open(path, "r") as fobj:
+            lock = json.load(fobj)
+        lock = SCHEMA(convert_to_unicode(lock))
+    except FileNotFoundError:
+        lock = SCHEMA({})
+    except JSONDecodeError as exc:
+        raise RWLockFileCorruptedError(path, cause=exc)
+    except Invalid as exc:
+        raise RWLockFileFormatError(path, cause=exc)
+    lock = defaultdict(dict, lock)
+    lock["read"] = defaultdict(list, lock["read"])
+    lock["write"] = defaultdict(dict, lock["write"])
+    yield lock
+    with open(path, "w+") as fobj:
+        json.dump(lock, fobj)
+
+
+def _info_to_str(info):
+    return "  (PID {}): {}".format(info["pid"], info["cmd"])
+
+
+def _get_urls(lock, path_info):
+    ret = set(filter(path_info.overlaps, lock.keys()))
+    ret.add(path_info.url)
+    return ret
+
+
+def _check_no_writers(lock, info, path_infos):
+    for path_info in path_infos:
+        urls = _get_urls(lock["write"], path_info)
+        writers = [lock["write"].get(url) for url in urls]
+        writers = list(filter(bool, writers))
+        if not writers:
+            continue
+
+        assert len(writers) == 1
```

Comment:
Again there might be several blocking writers. Say we try to read lock "/a/b" and there are:
```python
{"write":
    {"/a/b/foo": {"pid": 1001, ...}}, 
    {"/a/b/bar": {"pid": 1002, ...}}, 
}
```

Reply:
Great point! Haven't noticed that. Fixing.